### PR TITLE
Converted ckeditor5-dev-* packages to ESM

### PR DESCRIPTION
**Pull request by pomek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix (release-tools): Passing an odd number as a value of the `concurrency` option will not crash the `executeInParallel()` function. Closes ckeditor/ckeditor5#17025.

Other: The CKEditor 5 Dev packages are now ESM.

Other: The `@ckeditor/jsdoc-plugins` package is no longer available as CKEditor 5 documentation uses TypeScript sources to prepare API docs. The package is no longer use by us anywhere. Hence, we decided to remove a dead code.

MAJOR BREAKING CHANGE (utils): The `builds.getDllPluginWebpackConfig()` function is now asynchronous now.

MAJOR BREAKING CHANGE (utils): The following functions from the `bundler` object are no longer available: `createEntryFile()`, `getEditorConfig()`, `getPlugins()`.

MAJOR BREAKING CHANGE (utils): The following functions from the `stream` object are no longer available: `isTestFile()`, `isSourceFile()`, `isJSFile()`.

MAJOR BREAKING CHANGE (utils): The `styles.themeLogger()` function is no longer exposed publicly.

MAJOR BREAKING CHANGE (tests): The IntelliJ Karma configuration path needs to be updated. Now, it ends with the `cjs` suffix instead of `js`.

MAJOR BREAKING CHANGE (docs): The package uses only TypeDoc to build the documentation. The `build()` function no longer supports `type` property in the configuration.

MAJOR BREAKING CHANGE (docs): An array of ignored files must be specified as `ignoreFiles` instead of negating a glob pattern in `sourceFiles`.

MAJOR BREAKING CHANGE: The following packages have been converted to ESM. To load them, use an `import` statement instead of `require()`.
  * [`@ckeditor/ckeditor5-dev-build-tools`](https://www.npmjs.com/package/@ckeditor/ckeditor5-dev-build-tools)
  * [`@ckeditor/ckeditor5-dev-bump-year`](https://www.npmjs.com/package/@ckeditor/ckeditor5-dev-bump-year)
  * [`@ckeditor/ckeditor5-dev-ci`](https://www.npmjs.com/package/@ckeditor/ckeditor5-dev-ci)
  * [`@ckeditor/ckeditor5-dev-dependency-checker`](https://www.npmjs.com/package/@ckeditor/ckeditor5-dev-dependency-checker)
  * [`@ckeditor/ckeditor5-dev-docs`](https://www.npmjs.com/package/@ckeditor/ckeditor5-dev-docs)
  * [`@ckeditor/ckeditor5-dev-release-tools`](https://www.npmjs.com/package/@ckeditor/ckeditor5-dev-release-tools)
  * [`@ckeditor/ckeditor5-dev-stale-bot`](https://www.npmjs.com/package/@ckeditor/ckeditor5-dev-stale-bot)
  * [`@ckeditor/ckeditor5-dev-tests`](https://www.npmjs.com/package/@ckeditor/ckeditor5-dev-tests)
  * [`@ckeditor/ckeditor5-dev-transifex`](https://www.npmjs.com/package/@ckeditor/ckeditor5-dev-transifex)
  * [`@ckeditor/ckeditor5-dev-translations`](https://www.npmjs.com/package/@ckeditor/ckeditor5-dev-translations)
  * [`@ckeditor/ckeditor5-dev-utils`](https://www.npmjs.com/package/@ckeditor/ckeditor5-dev-utils)
  * [`@ckeditor/ckeditor5-dev-web-crawler`](https://www.npmjs.com/package/@ckeditor/ckeditor5-dev-web-crawler)

---

### Additional information

Resolve the following TODOs before merging:

* [x] https://github.com/ckeditor/ckeditor5-dev/pull/1011/files/890186e9189c5b9610441a509e73ce8daff2ff9f#diff-e2954b558f2aa82baff0e30964490d12942e0e251c1aa56c3294de6ec67b7cf5
* [x] https://github.com/ckeditor/ckeditor5-dev/pull/1011/files/890186e9189c5b9610441a509e73ce8daff2ff9f#diff-e2954b558f2aa82baff0e30964490d12942e0e251c1aa56c3294de6ec67b7cf5
* [x] https://github.com/ckeditor/ckeditor5-dev/pull/1011/files/890186e9189c5b9610441a509e73ce8daff2ff9f#diff-e2954b558f2aa82baff0e30964490d12942e0e251c1aa56c3294de6ec67b7cf5
* [x] https://github.com/ckeditor/ckeditor5-dev/pull/1011#pullrequestreview-2314408021
* [x] https://github.com/ckeditor/ckeditor5-dev/pull/1011#discussion_r1768356461
* [x] https://github.com/ckeditor/ckeditor5-dev/pull/1011#discussion_r1771187171

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -13,11 +13,17 @@ module.exports = {
 	ignorePatterns: [
 		'**/dist/*',
 		'**/coverage/**',
-		'**/node_modules/**'
+		'**/node_modules/**',
+
+		// ESLint does not understand `import ... with { ... }`.
+		// See: https://github.com/eslint/eslint/discussions/15305.
```
Comment (pomek) [inline .eslintrc.js:19]: TODO: To resolve.

File: .eslintrc.js [hunk 2]
```diff
@@ -13,11 +13,17 @@ module.exports = {
...
+		'packages/ckeditor5-dev-ci/lib/data/index.js',
+		'packages/ckeditor5-dev-transifex/lib/data/index.js'
 	],
 	rules: {
 		'no-console': 'off',
 		'ckeditor5-rules/require-file-extensions-in-imports': 'off',
+		'mocha/no-global-tests': 'off', // TODO: remove when all mocha tests are removed.
```
Comment (pomek) [inline .eslintrc.js:26]: TODO: To remove.

File: .eslintrc.js [hunk 3]
```diff
@@ -29,9 +35,27 @@ module.exports = {
 	},
 	overrides: [
 		{
-			files: [ './packages/ckeditor5-dev-build-tools/tests/**/*' ],
+			files: [
+				// TODO: add packages as they are migrated to ESM.
```
Comment (pomek) [inline .eslintrc.js:39]: TODO: To resolve.

**Review by pomek — COMMENTED** [review]
TODO: Review JSDoc types.

File: .circleci/template.yml [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+          name: Prepare environment variables
+          command: |
+            #!/bin/bash
+
+            # Non-secret environment variables needed for the pipeline scripts.
+            CKE5_GITHUB_ORGANIZATION="ckeditor"
+            CKE5_GITHUB_REPOSITORY="ckeditor5-dev"
+            CKE5_CIRCLE_APPROVAL_JOB_NAME="release_approval"
+            # CKE5_GITHUB_RELEASE_BRANCH="master"
+            CKE5_GITHUB_RELEASE_BRANCH="epic/3757-esm"
```
Comment (pomek) [inline .circleci/template.yml:60]: This change must be updated to `#master` before merging. The entire commit 1cc2de522b9771463d32f9453f18814f87fbca24 should be reverted.

File: scripts/changelog.js [hunk 5]
```diff
@@ -5,22 +5,17 @@
...
-const parseArguments = require( './utils/parsearguments' );
+import { generateChangelogForMonoRepository } from '@ckeditor/ckeditor5-dev-release-tools';
+import parseArguments from './utils/parsearguments.js';
 
 const cliArguments = parseArguments( process.argv.slice( 2 ) );
 
 const changelogOptions = {
 	cwd: process.cwd(),
 	packages: 'packages',
+	mainBranch: 'epic/3757-esm',
```
Comment (pomek) [inline scripts/changelog.js:16]:
```suggestion
```

To revert before merging.

File: .eslintrc.js (hunk 1 above)
Comment (pomek) [reply .eslintrc.js]: It turns out that you need to install additional parsers that may conflict with TypeScript. Let's leave it as it is.

File: .eslintrc.js (hunk 2 above)
Comment (pomek) [reply .eslintrc.js]: e567631

File: .eslintrc.js (hunk 3 above)
Comment (pomek) [reply .eslintrc.js]: e567631

**Comment by pomek** [comment]
I think the job is done. I do not want to revert the alpha release changes if we want to re-release it.

When it comes to merging the PR, I would like to handle it once CKEditor 5 releases its next minor.

File: .circleci/template.yml (hunk 4 above)
Comment (pomek) [reply .circleci/template.yml]: Done in 2722abce9c964cc1a800af51ab5cd54510db12ab.

**Review by psmyrek — APPROVED** [review]
LGTM 🎉

**Pull request merged**